File: 1512. Number of Good Pairs.java
/**
 * https://leetcode.com/problems/number-of-good-pairs/
 */

public class NumberOfGoodPairs {
    // O(n) time complexity
    // O(101) space complexity
    int numIdenticalPairs(int[] nums) {
        int[] arr = new int[101];
        for (int num : nums)
            arr[num]++;
        int res = 0;
        for (int ar : arr) {
            res += ar > 1 ? ar * (ar - 1) / 2 : 0;
        }
        return res;
    }
}

class Main {
    public static void main(String[] args) {
        NumberOfGoodPairs n = new NumberOfGoodPairs();
        int[] nums = {1, 2, 3, 1, 1, 3};
        System.out.println(n.numIdenticalPairs(nums)); // Output: 4
    }
}

File: Kotlin/1512. Number of Good Pairs.kt
/**
 * https://leetcode.com/problems/number-of-good-pairs/
 */

// O(n) time complexity
// O(101) space complexity
fun numIdenticalPairs(nums: IntArray): Int {
    val arr = IntArray(101)
    nums.forEach {
        arr[it]++
    }
    var res = 0
    arr.forEach {
        res += if (it > 1) it * (it - 1) / 2 else 0
    }
    return res
}

fun main() {
    val nums = intArrayOf(1, 2, 3, 1, 1, 3)
    println(numIdenticalPairs(nums)) // Output: 4

}